### PR TITLE
Query generator from pr 581

### PR DESCRIPTION
**Pull request by czaloom**
### Problem Description

The existing query generator will not be able to handle nested logic in filters which is an upcoming feature in PR #581.

This PR also implements the new filtering schemas from PR #581 under `valor_api/schemas/filters.py` but does not expose them at the API level.

### Feature Description

The new generator separates the solving of individual filter expressions into CTE's which are then joined using logical expressions (AND, OR, NOT) to the main query.

As GroundTruths and Predictions are stored in the same `Annotation` table there is a need to define different JOIN sequences to facilitate all possible queries.

## GroundTruth as Label Source

Useful for returning groundtruths.

Q: Get me ground truth labels that have also gotten a score greater than 0.5 from Model A.

```mermaid
graph TD
    B[Model] --- C[Datum]
    A[Dataset] --- C
    C --- D[Annotation]
    D --- E[GroundTruth]
    E --- F[Label]
    F --- G[Prediction]
    D --- H[Embedding]
```

## Prediction as Label Source

Useful for returning predictions.

Q: Get me predictions that have a label == (k1, v1) over dataset B.

```mermaid
graph TD
    G[Model] --- C[Annotation]
    A[Dataset] --- B[Datum]
    B --- C
    C --- D[Prediction]
    D --- E[Label]
    E --- F[GroundTruth]
    C --- H[Embedding]
```

## Annotation as Label Source

Useful for returning sets of annotations regardless of whether they are groundtruths or predictions.

Q: Get the set of annotations (gts + pds) that operate over labels that have score > 3 from Model A.

```mermaid
graph TD
    A[Model] --- D[Annotation]
    B[Dataset] --- C[Datum]
    C --- D[Annotation]
    D --- E[Label]
    E --- F[GroundTruth]
    E --- G[Prediction]
    D --- H[Embedding]
```

### CODE REVIEW

File: api/tests/functional-tests/backend/metrics/test_classification.py [hunk 1]
```diff
@@ -597,6 +597,10 @@ def test_compute_roc_auc_groupby_metadata(
         dataset_names=[dataset_name],
         task_types=[enums.TaskType.CLASSIFICATION],
         datum_metadata={"md1": [schemas.StringFilter(value="md1-val0")]},
+        labels=[
```
Comment (ntlind) [inline api/tests/functional-tests/backend/metrics/test_classification.py:600]: nit: does this label do anything?

File: api/tests/functional-tests/backend/query/test_ops.py [hunk 2]
```diff
@@ -653,102 +676,157 @@ def model_sim(
...
     assert (model_name1,) in model_names
     assert (model_name2,) in model_names
 
-    # Q: Get names for models where label=tree exists in predictions
-    f = schemas.Filter(labels=[{"class": "tree"}])
-    q = Query(models.Model.name).filter(f).predictions()
-    model_names = db.query(q).distinct().all()  # type: ignore - SQLAlchemy type issue
+    # Q: Get names for models that operate over dataset that doesn't exist.
+    f = Filter(predictions=create_dataset_filter("invalid"))
+    model_names = select_from_annotations(
```
Comment (ntlind) [inline api/tests/functional-tests/backend/query/test_ops.py:769]: nit: it'd be easier to follow these tests if they were split up by groundtruths / prediction / etc. (i.e., `test_select_from_predictions`, `test_select_from_groundtruths`, etc.)

Comment (ntlind) [reply api/tests/functional-tests/backend/query/test_ops.py:769]: the overall test coverage here seems solid, so I'm not too worried about this one

File: api/tests/functional-tests/backend/query/test_ops.py [hunk 3]
```diff
@@ -1033,50 +1224,90 @@ def test_multiple_tables_in_args(
     ) in name_pairings
 
 
+def create_geospatial_inside_filter(
+    symbol_name: str,
+    value: Value,
```
Comment (ntlind) [inline api/tests/functional-tests/backend/query/test_ops.py:1229]: nit: `value` is very generic; consider making this code easier to follow by using a more specific name

File: api/tests/functional-tests/backend/query/test_ops.py [hunk 4]
```diff
@@ -1033,50 +1224,90 @@ def test_multiple_tables_in_args(
...
 def _get_geospatial_names_from_filter(
     db: Session,
-    geodict: dict[
-        str,
-        list[list[list[list[float | int]]]]
-        | list[list[list[float | int]]]
-        | list[float | int]
-        | str,
-    ],
+    value: Value,
```
Comment (ntlind) [inline api/tests/functional-tests/backend/query/test_ops.py:1271]: nit: `value` is very generic; consider making this code easier to follow by using a more specific name

File: api/valor_api/backend/query/filtering.py [hunk 5]
```diff
@@ -1,430 +1,379 @@
...
-    return ops[opstr]
-
-
-def _get_spatial_op(opstr) -> Callable:
-    """Returns function if operator is valid for spatial comparison."""
-    ops = {
-        "intersect": lambda lhs, rhs: func.ST_Intersects(lhs, rhs),
-        "inside": lambda lhs, rhs: func.ST_Covers(rhs, lhs),
-        "outside": lambda lhs, rhs: not_(func.ST_Covers(rhs, lhs)),
+opstr_to_operator = {
```
Comment (ntlind) [inline api/valor_api/backend/query/filtering.py:49]: nit: add short docstrings above these dictionaries describing what they're used for?

File: api/valor_api/backend/query/filtering.py [hunk 6]
```diff
@@ -1,430 +1,379 @@
...
+            tables.add(Embedding)
+    return tables
+
+
+def generate_dependencies(
+    func: OneArgFunction | TwoArgFunction | NArgFunction | None,
+) -> tuple[int | dict | None, list[CTE], list[TableTypeAlias]]:
     """
-    Constructs sqlalchemy expressions using label filters.
+    Recursively generates the dependencies for createing a filter subquery.
```
Comment (ntlind) [inline api/valor_api/backend/query/filtering.py:305]:
```suggestion
    Recursively generates the dependencies for creating a filter subquery.
```

File: api/valor_api/backend/query/mapping.py [hunk 7]
```diff
@@ -69,14 +67,10 @@ def _recursive_select_to_table_names(
...
         return [argument.name]
-    if isinstance(argument, TableTypeAlias):
-        return [argument.__tablename__]
-    if isinstance(argument, DeclarativeMeta):
-        if "__tablename__" not in argument.__dict__:
-            raise AttributeError(
-                f"DeclarativeMeta object '{argument}' missing __tablename__ attribute."
-            )
-        return [argument.__dict__["__tablename__"]]
+    # elif isinstance(argument, TableTypeAlias):
```
Comment (ntlind) [inline api/valor_api/backend/query/mapping.py:70]: remove commented code?

File: api/valor_api/backend/query/mapping.py [hunk 8]
```diff
@@ -106,7 +100,7 @@ def _recursive_select_to_table_names(
         )
 
 
-def map_arguments_to_tables(args: tuple[Any]) -> set[TableTypeAlias]:
+def map_arguments_to_tables(*args) -> set[TableTypeAlias]:
```
Comment (ntlind) [inline api/valor_api/backend/query/mapping.py:103]: add back typing to args? args are still defined in the docstring (tuple[Any])

File: api/valor_api/backend/query/ops.py [hunk 9]
```diff
@@ -1,26 +1,100 @@
...
+    *args, filter_: AdvancedFilter | None = None
+) -> Select[Any]:
+    """
+    Creates a select statement from provided arguments and filters.
+
+    Labels are queried from the combination of groundtruths and predictions.
+
+    Parameters
+    ----------
+    *args : tuple[Any]
```
Comment (ntlind) [inline api/valor_api/backend/query/ops.py:20]: args are still defined in docstring. can we add them back to the parameter?

File: api/valor_api/backend/query/solvers.py [hunk 10]
```diff
@@ -1,442 +1,690 @@
...
     """
-    Selects best fitting graph to run query generation and returns tuple(query, subquery | None).
+    Generates the main query.
 
-    Description
+    Includes all args-related and filter-related tables.
+
+    Parameters
     ----------
-    To construct complex queries it is necessary to describe the relationship between predictions and groundtruths.
```
Comment (ntlind) [inline api/valor_api/backend/query/solvers.py:334]: nit: could we update and add back these longer docstrings somewhere? I found them helpful

File: api/valor_api/schemas/filters.py [hunk 11]
```diff
@@ -11,6 +11,308 @@
...
+
+
+class Symbol(BaseModel):
+    type: str
+    name: str
+    key: str | None = None
+    attribute: str | None = None
+
+
+class Value(BaseModel):
```
Comment (ntlind) [inline api/valor_api/schemas/filters.py:69]: add docstrings to these classes?

File: api/valor_api/schemas/filters.py [hunk 12]
```diff
@@ -46,6 +348,21 @@ def _validate_comparison_operator(cls, op: str) -> str:
 
     model_config = ConfigDict(extra="forbid")
 
+    def to_function(
```
Comment (ntlind) [inline api/valor_api/schemas/filters.py:351]: add docstring describing what `to_function` is used for?

File: api/tests/functional-tests/backend/query/test_ops.py (hunk 3 above)
Comment (czaloom) [reply api/tests/functional-tests/backend/query/test_ops.py:1229]: Its meant to be generic because it represents literally any value that is not symbolic.

Comment (czaloom) [reply api/tests/functional-tests/backend/query/test_ops.py:1229]: Do you have any suggestions?

File: api/tests/functional-tests/backend/query/test_ops.py (hunk 4 above)
Comment (czaloom) [reply api/tests/functional-tests/backend/query/test_ops.py:1271]: See above. Resolving this as it is a duplicate.

File: api/valor_api/backend/query/mapping.py (hunk 8 above)
Comment (czaloom) [reply api/valor_api/backend/query/mapping.py:103]: *args is python magic and defines a variable number of inputs that will be formatted into a ordered tuple.

Ill update the docstring to say the following:

```
    Parameters
    ----------
    *args : tuple[Any]
        A variable length list of arguments from a selection statement.
```

Comment (czaloom) [reply api/valor_api/backend/query/mapping.py:103]: If you define the function with `def fn(args: tuple[Any])` the input has to be passed as a preformatted tuple.

File: api/tests/functional-tests/backend/query/test_ops.py (hunk 2 above)
Comment (czaloom) [reply api/tests/functional-tests/backend/query/test_ops.py:769]: Ill make a note of this for the integration tests in the next pr

File: api/tests/functional-tests/backend/query/test_ops.py (hunk 3 above)
Comment (ntlind) [reply api/tests/functional-tests/backend/query/test_ops.py:1229]: `inputted_value`, `user_input`, etc.; some other name which provides context about what this "value" means

**Pull request merged**